### PR TITLE
Add RelayOrderExecutor and unit tests

### PR DESCRIPTION
**Pull request by zhongeric**

### CODE REVIEW

File: src/sample-executors/RelayOrderExecutor.sol [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+    modifier onlyWhitelistedCaller() {
+        if (msg.sender != whitelistedCaller) {
+            revert CallerNotWhitelisted();
+        }
+        _;
+    }
+
+    constructor(address _whitelistedCaller, IRelayOrderReactor _reactor, address _owner) Owned(_owner) {
+        whitelistedCaller = _whitelistedCaller;
```
Comment (snreynolds) [inline src/sample-executors/RelayOrderExecutor.sol:32]: We should probably allow changing the whitelistedCaller? Or allow execute/executBatch to be callable by the owner as well

File: test/foundry-tests/executors/RelayOrderExecutorTest.t.sol [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+        filler = vm.addr(fillerPrivateKey);
+        executorOwner = vm.addr(executorOwnerPrivateKey);
+
+        permit2 = IPermit2(deployPermit2());
+
+        reactor = new RelayOrderReactor(permit2);
+        executor = new RelayOrderExecutor(filler, IRelayOrderReactor(reactor), executorOwner);
+    }
+
+    // Test:
```
Comment (snreynolds) [inline test/foundry-tests/executors/RelayOrderExecutorTest.t.sol:49]: TODO?

Comment (snreynolds) [reply test/foundry-tests/executors/RelayOrderExecutorTest.t.sol:49]: Would add some edge case tests like caller not whitelisted does a multicall permit and execute and it fails on execute

File: test/foundry-tests/reactors/RelayOrderReactorTest.t.sol [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+        reactor = new RelayOrderReactor(permit2);
+        executor = new RelayOrderExecutor(filler, IRelayOrderReactor(reactor), executorOwner);
+
+        // swapper approves permit2 to transfer tokens
+        tokenIn.forceApprove(swapper, address(permit2), type(uint256).max);
+        assertEq(tokenIn.allowance(swapper, address(permit2)), type(uint256).max);
+    }
+
+    /// @notice Create and return a basic single Relay order along with its signature, orderHash, and orderInfo
+    function createAndSignOrder(RelayOrder memory order)
```
Comment (snreynolds) [inline test/foundry-tests/reactors/RelayOrderReactorTest.t.sol:63]: not used anywhere?

Comment (snreynolds) [reply test/foundry-tests/reactors/RelayOrderReactorTest.t.sol:63]: also probably candidate to put in a utils file or something if we do end up wanting something like this

File: src/sample-executors/RelayOrderExecutor.sol (hunk 1 above)
Comment (zhongeric) [reply src/sample-executors/RelayOrderExecutor.sol:32]: Yea so in the executor we use in prod for UniswapX today we actually allow for any of 10 whitelisted callers. That was mostly a requirement for scaling - but can consider a better design here.

Comment (zhongeric) [reply src/sample-executors/RelayOrderExecutor.sol:32]: Let's just allow for owner to change caller in this version and can change if needed

File: test/foundry-tests/executors/RelayOrderExecutorTest.t.sol (hunk 2 above)
Comment (zhongeric) [reply test/foundry-tests/executors/RelayOrderExecutorTest.t.sol]: Yep added - did a test for multicall where caller has all perms needed, mixed perms, and no perms

File: test/foundry-tests/reactors/RelayOrderReactorTest.t.sol (hunk 3 above)
Comment (zhongeric) [reply test/foundry-tests/reactors/RelayOrderReactorTest.t.sol:63]: Going to remove, this was more helpful for X tests

File: src/sample-executors/RelayOrderExecutor.sol [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+    /// @notice Transfer all ETH in this contract to the recipient. Can only be called by owner.
+    /// @param recipient The recipient of the ETH
+    function withdrawETH(address recipient) external onlyOwner {
+        SafeTransferLib.safeTransferETH(recipient, address(this).balance);
+    }
+
+    /// @notice Change the whitelisted caller
+    /// @param newWhitelistedCaller The new whitelisted caller
+    function changeWhitelistedCaller(address newWhitelistedCaller) external onlyOwner {
+        whitelistedCaller = newWhitelistedCaller;
```
Comment (snreynolds) [inline src/sample-executors/RelayOrderExecutor.sol:77]: Maybe revert if its the same as the current whitelistedCaller?

File: src/sample-executors/RelayOrderExecutor.sol [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+        emit WhitelistedCallerChanged(newWhitelistedCaller);
+    }
+
+    receive() external payable {}
```
Comment (snreynolds) [inline src/sample-executors/RelayOrderExecutor.sol:81]: there really should never be eth in this contract no?

**Review by snreynolds — COMMENTED** [review]
lgtm! Although some point, we should probably add full end to end integration, with a permit and actual relay order. wdyt?

File: src/sample-executors/RelayOrderExecutor.sol (hunk 4 above)
Comment (zhongeric) [reply src/sample-executors/RelayOrderExecutor.sol:77]: Don't think its necessary since solmate doesn't do that either for ownership... not really opinionated about it lmk what you think

Comment (snreynolds) [reply src/sample-executors/RelayOrderExecutor.sol:70]: sure sg

File: src/sample-executors/RelayOrderExecutor.sol (hunk 5 above)
Comment (snreynolds) [reply src/sample-executors/RelayOrderExecutor.sol:74]: makes sense

File: src/sample-executors/RelayOrderExecutor.sol [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+    }
+
+    constructor(address _whitelistedCaller, IRelayOrderReactor _reactor, address _owner) Owned(_owner) {
+        whitelistedCaller = _whitelistedCaller;
+        reactor = _reactor;
+    }
+
+    /// @notice Shortcut to execute a single relay order
+    /// @param order The order to execute
+    function execute(SignedOrder calldata order) external onlyWhitelistedCaller {
```
Comment (snreynolds) [inline src/sample-executors/RelayOrderExecutor.sol:39]: I think that with multicall existing ON the reactor we really dont need these wrapper functions and this contract isn't really an executor anymore but a fee controller... no?

**Review by snreynolds — CHANGES_REQUESTED** [review]
Basically I think that the flow would just be to set the recipient of the fees to this contract and this contract sole purpose is to collect those... but the actual call to reactor would just be made directly since it has multicall now

thoughts?

And we don't even need it to be "executor" anymore

**Comment by zhongeric** [comment]
Yes I think practically we'd do that - I guess it's just do we want a "sample" contract for integrators to reference for how to integrate or move that into a README since its pretty straightforward.

**Comment by snreynolds** [comment]
> Yes I think practically we'd do that - I guess it's just do we want a "sample" contract for integrators to reference for how to integrate or move that into a README since its pretty straightforward.

Yeah we could a have a sample if you want. That probably doesnt need to have any of the ownership stuff cause thats really specific to our use case. But I think regardless we want a contract that is more of our FeeRecipient

**Review by snreynolds — APPROVED** [review]
approving, @zhongeric chatted offline will edit this to be a sample contract.. since we are adding fee recipient on reactor
